### ISSUE TITLE
Overcoming Catastrophic Forgetting in Incremental Few-Shot Learning by Finding Flat Minima

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#144]
## TL; DR

- Find flat minima in base training, and restrict parameter updates in incremental few-shot training phase.

## Motivation

- Incremental Learning + Few-shot = ***little gain*** from plasticity, more to lose. (data scarcity)
- SOTA Incremental Few-shot Learning (IFL) methods lag behind the simple baseline of freezing representations from base training and applying prototype nearest neighbor classification in terms of performance.
- Hence, preserving knowledge from base training is important (stability > plasticity) while working on backward transfer (BWT) is not enough.

## Prior Works

- Regularization-based methods

## Method

![image](https://user-images.githubusercontent.com/46308407/147867715-33f1d09b-1fb5-4d1a-b152-9b9b0e9f2b48.png)

- Baseline refers to *freezing* the representation from **base** training, and predicting new classes based on distance from class prototype vectors.
- Since SOTA IFL methods show inferior performance to such naive baseline, we can conclude that preserving knowledge from base classes is more crucial than learning something new from few-shot new classes.
- Also, regularization-based methods are usually insufficient to combat catastrophic forgetting.

![image](https://user-images.githubusercontent.com/46308407/147867738-ce690af8-9ec0-4e9d-baf7-b28f41d26ebb.png)

- Idea is similar to Sharpness-Aware Minimization (SAM, ICLR'21).
- Find flat local minima from base training, and clamp parameter updates in later incremental few-shot training stage, so that model parameters stay in the flat region.
- Similar to freezing feature extractor in effect, but a little more flexible as the model looks for robust minima in the base training stage as well.

### Finding Flat Minima (F2M)

- Inject noise to conv layer feature extractor parameters and minimize the conventional cross-entropy objective.
- Hence, the model seeks to find region in the parameter space where small perturbation doesn't affect loss value much. (=flat)

### Preserving Knowledge

- When fine-tuning in the incremental few-shot stage, simply ***clamp*** the parameters so that they stay in the flat region.

### Objective and Algorithm

![image](https://user-images.githubusercontent.com/46308407/147867784-f0ea2c14-59ee-4489-9419-a8f8ddd54f87.png)

- Inject noise *M* times.
- Regularize prototype drift (change in class prototype vectors) with simple L2 loss term added on the right.

![image](https://user-images.githubusercontent.com/46308407/147867797-3926d05a-45c0-4767-8917-127904b0684c.png)

- For fine-tuning, resort to metric-based classification based on Euclidean distance. (Distance from each class's prototype)

![image](https://user-images.githubusercontent.com/46308407/147867805-dae29a88-e7d7-43db-9fc4-1e5580c08a50.png)

- Convergence analysis is also provided based on relatively mild assumptions (L-smooth objective, learning rate bound, etc)

### Comparison with Baselines

![image](https://user-images.githubusercontent.com/46308407/147867811-a5df551f-19f3-426f-9dd7-e071fb85c017.png)
![image](https://user-images.githubusercontent.com/46308407/147867814-72c1413f-14ba-492e-8969-1daf7ae388ec.png)

- M=2~4
- b = 0.01 (clamping hyperparam)
- Epoch=6, lr=0.02

### Flatness

![image](https://user-images.githubusercontent.com/46308407/147867838-612b25a3-638c-48a0-a5d3-3bf64d618338.png)

- Add random noises to the obtained minima, and compute deviations. (Flat -> small changes around the minima)

### Ablation

![image](https://user-images.githubusercontent.com/46308407/147867853-41d34262-4bf7-482b-aa5c-eddd832452ac.png)

- Validate effectiveness of each components.
- Note that gains from clamping is huge, which is the crucial part of their work. (also indicates that preserving knowledge is the key in IFL)

## Comments

- Thinking about landscape and the geometry is always intriguing.
- Overall idea is quite similar to SAM (ICLR'21), but the formulation differs. Authors here simply inject noise in the optimization process.
- I do get the impression that IFL is a research area that is quite *under-explored*, so although the result obtained by this work is not that impressive it still outperforms previous methods. (freezing baseline being the SOTA tells us a lot)
- Idea is interesting and method is simply, all that is good. However, still I think we should be able to step a little further than sitting there and fearing to lose what we've learned previously. I know it's very challenging especially in few-shot setting, but still in the future I expect brilliant works that take ambitious steps ahead, to actively reason and learn new concepts even from limited data.
